src/handlers/miner_hbbft_handler.erl

```diff
@@ -93,7 +95,30 @@ metadata(Version, Meta, Chain) ->
                                 lager:info("no snapshot interval configured"),
                                 ChainMeta0
                         end,
-            t2b(maps:merge(Meta, ChainMeta))
+            ChainMeta1 =
+                case blockchain:config(?poc_challenger_type, Ledger) of
+                    {ok, validator} ->
+                        %% get a set of ephemeral keys for POC usage
+                        %% keys come from the proposals cache
+                        %% populated by validators via key proposals
+                        %% in their heartbeats
+                        ChallengeRate =
+                            case blockchain:config(?poc_challenge_rate, Ledger) of
+                                {ok, CR} -> CR;
+                                _ -> 1
+                            end,
+                        lager:debug("poc challenge rate ~p", [ChallengeRate] ),
+                        SelfPubKeyBin = blockchain_swarm:pubkey_bin(),
+                        NumKeys = max(1, trunc(ChallengeRate / (((N-1)/3) * 2 ))),
+                        POCEphemeralKeys = miner_poc_mgr:get_random_poc_key_proposals(NumKeys, Ledger),
```

Comment:
It's worth thinking about the determinism or accountability here, is there a way to prove this node could not bias its choices here if it was trying to play favorites?

Reply:
We should discuss this directly.